### PR TITLE
Do not add nulls to the hash table when null_equality::NOT_EQUAL is passed to left_semi_join and left_anti_join

### PR DESCRIPTION
**Pull request by nvdbaranec**
Fixes  https://github.com/rapidsai/cudf/issues/7300

This is fundamentally the same issue and fix as https://github.com/rapidsai/cudf/pull/6943/files from @hyperbolic2346 

When nulls are considered not equal (`null_equality::NOT_EQUAL`) there is no point in adding them to the hash table used for the join as they will never compare as true against anything.  Adding large numbers of nulls was causing huge performance issues.

Includes a fix to doxygen comments for `left_anti_join`

Performance gain is tremendous.

Before:

```
Benchmark                                                                             Time             CPU   Iterations
-----------------------------------------------------------------------------------------------------------------------
Join<int32_t, int32_t>/left_anti_join_32bit_nulls/100000/100000/manual_time        1072 ms         1072 ms            1
Join<int32_t, int32_t>/left_anti_join_32bit_nulls/200000/400000/manual_time        4253 ms         4253 ms            1
Join<int32_t, int32_t>/left_anti_join_32bit_nulls/300000/1000000/manual_time      14016 ms        14016 ms            1
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/100000/100000/manual_time         932 ms          932 ms            1
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/200000/400000/manual_time        4481 ms         4481 ms            1
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/300000/1000000/manual_time      14172 ms        14172 ms            1
```

After:
```
-----------------------------------------------------------------------------------------------------------------------
Benchmark                                                                             Time             CPU   Iterations
-----------------------------------------------------------------------------------------------------------------------
Join<int32_t, int32_t>/left_anti_join_32bit_nulls/100000/100000/manual_time       0.143 ms        0.162 ms         4996
Join<int32_t, int32_t>/left_anti_join_32bit_nulls/200000/400000/manual_time       0.255 ms        0.275 ms         2780
Join<int32_t, int32_t>/left_anti_join_32bit_nulls/300000/1000000/manual_time      0.514 ms        0.532 ms         1368
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/100000/100000/manual_time       0.135 ms        0.155 ms         5203
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/200000/400000/manual_time       0.206 ms        0.224 ms         3325
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/300000/1000000/manual_time      0.368 ms        0.385 ms         1903
```

### CODE REVIEW

**Comment by nvdbaranec** [comment]
I noticed as I was doing this that we have no tests for left_anti_join.  Will add.

File: cpp/src/join/semi_join.cu [hunk 1]
```diff
@@ -92,12 +92,22 @@ std::unique_ptr<rmm::device_uvector<cudf::size_type>> left_semi_anti_join(
...
   auto hash_table     = *hash_table_ptr;
 
-  thrust::for_each_n(rmm::exec_policy(stream),
-                     thrust::make_counting_iterator<size_type>(0),
-                     right_num_rows,
-                     [hash_table] __device__(size_type idx) mutable {
-                       hash_table.insert(thrust::make_pair(idx, true));
-                     });
+  // if compare_nulls == NOT_EQUAL, we can simply ignore any rows that are
+  // entirely NULL as they will never compare to equal.
```
Comment (hyperbolic2346) [inline cpp/src/join/semi_join.cu:96]: This isn't exactly true is it? The row needs a single null entry to skip the hash table. It isn't required that the row is entirely null.

I have a feeling this is a copy/paste thing and the original comment is incorrect as well.

File: cpp/src/join/semi_join.cu [hunk 2]
```diff
@@ -92,12 +92,22 @@ std::unique_ptr<rmm::device_uvector<cudf::size_type>> left_semi_anti_join(
...
+                             ? rmm::device_buffer{0, stream}
+                             : cudf::detail::bitmask_and(right_flattened_keys, stream);
+  // skip rows that are null here.
+  thrust::for_each_n(
+    rmm::exec_policy(stream),
+    thrust::make_counting_iterator<size_type>(0),
+    right_num_rows,
+    [hash_table, row_bitmask = static_cast<bitmask_type const*>(row_bitmask.data())] __device__(
+      size_type idx) mutable {
+      if (!row_bitmask || cudf::bit_is_set(row_bitmask, idx)) {
```
Comment (harrism) [inline cpp/src/join/semi_join.cu:107]: Doesn't `bit_is_set` already check that row_bitmask is not null?

File: cpp/benchmarks/join/join_benchmark.cu [hunk 3]
```diff
@@ -37,8 +37,8 @@ template <typename key_type, typename payload_type>
 class Join : public cudf::benchmark {
 };
 
-template <typename key_type, typename payload_type, bool Nullable>
-static void BM_join(benchmark::State &state)
+template <typename key_type, typename payload_type, bool Nullable, typename Join>
+static void BM_join(Join JoinFunc, benchmark::State& state)
```
Comment (harrism) [inline cpp/benchmarks/join/join_benchmark.cu:41]: I think other benchmarks make `state` the first param. Could be wrong.

**Review by harrism — CHANGES_REQUESTED** [review]
Nice speedup!

File: cpp/src/join/semi_join.cu [hunk 4]
```diff
@@ -92,12 +92,22 @@ std::unique_ptr<rmm::device_uvector<cudf::size_type>> left_semi_anti_join(
                                                 equality_build);
   auto hash_table     = *hash_table_ptr;
 
-  thrust::for_each_n(rmm::exec_policy(stream),
-                     thrust::make_counting_iterator<size_type>(0),
-                     right_num_rows,
-                     [hash_table] __device__(size_type idx) mutable {
-                       hash_table.insert(thrust::make_pair(idx, true));
-                     });
+  // if compare_nulls == NOT_EQUAL, we can simply ignore any rows that are
```
Comment (vuule) [inline cpp/src/join/semi_join.cu:95]: Could we also adjust the hash table size based on the number of valid rows? Not sure how important this is for join operations, just a thought.

File: cpp/src/join/semi_join.cu (hunk 2 above)
Comment (nvdbaranec) [reply cpp/src/join/semi_join.cu:107]: It doesn't.  Interestingly, it has an actual assert.

File: cpp/src/join/semi_join.cu (hunk 4 above)
Comment (nvdbaranec) [reply cpp/src/join/semi_join.cu:95]: Something like this seems to work.  I have to do a `count_unset_bits()` on the aggregate bitmask though.  This also feels mildly spooky to change as I'm trying to treat lightly in this unfamiliar module :) 

```
auto const row_bitmask = (compare_nulls == null_equality::EQUAL)
                           ? rmm::device_buffer{}
                           : cudf::detail::bitmask_and(right_flattened_keys, stream);
size_type right_row_ignore_count = compare_nulls == null_equality::EQUAL ? 0 : 
  count_unset_bits(static_cast<const bitmask_type*>(row_bitmask.data()), 0, right_num_rows);
...
size_t const hash_table_size = compute_hash_table_size(right_num_rows - right_row_ignore_count);
```

**Review by jrhemstad — APPROVED** [review]

**Review by hyperbolic2346 — APPROVED** [review]

**Review by robertmaynard — APPROVED** [review]

**Review by harrism — APPROVED** [review]

**Comment by harrism** [comment]
@gpucibot merge

**Pull request merged**